include/dd/NoiseFunctionality.hpp

```diff
@@ -0,0 +1,204 @@
+
+            gateNoiseEffects = cGateNoiseEffects;
+            identityDD       = localDD->makeIdent(size);
+            localDD->incRef(identityDD);
+        }
+
+        dd::Qubit        nqubits;
+        double           noiseProbability = 0.0;
+        dd::ComplexValue sqrtAmplitudeDampingProbability{};
+        dd::ComplexValue oneMinusSqrtAmplitudeDampingProbability{};
+
+        double           noiseProbabilityMulti = 0.0;
+        dd::ComplexValue sqrtAmplitudeDampingProbabilityMulti{};
+        dd::ComplexValue oneMinusSqrtAmplitudeDampingProbabilityMulti{};
+
+        dd::GateMatrix ampDampingTrue      = {};
+        dd::GateMatrix ampDampingTrueMulti = {};
+
+        dd::GateMatrix ampDampingFalse      = {};
+        dd::GateMatrix ampDampingFalseMulti = {};
+
+        std::vector<dd::noiseOperations> gateNoiseEffects;
+
+        const std::unique_ptr<customPackage>& localDD;
+        size_t                                seed = 0;
+
+        dd::mEdge identityDD;
+
+        std::uniform_real_distribution<dd::fp> dist;
+        std::mt19937_64                        generator;
+
+        inline void applyNoiseOperation(const std::vector<dd::Qubit>& usedQubits, dd::mEdge dd_op, dd::vEdge& localRootEdge) {
+            bool multiQubitOperation = usedQubits.size() > 1;
+
+            for (auto& target: usedQubits) {
+                auto operation = generateNoiseOperation(dd_op, target, false, multiQubitOperation);
+                auto tmp       = localDD->multiply(operation, localRootEdge);
+
+                if (dd::ComplexNumbers::mag2(tmp.w) < dist(generator)) {
+                    operation = generateNoiseOperation(dd_op, target, true, multiQubitOperation);
+                    tmp       = localDD->multiply(operation, localRootEdge);
+                }
+
+                if (tmp.w != dd::Complex::one) {
+                    tmp.w = dd::Complex::one;
+                }
+                localDD->incRef(tmp);
+                localDD->decRef(localRootEdge);
+                localRootEdge = tmp;
+
+                // I only need to apply the operations once
+                dd_op = identityDD;
+            }
+        }
+
+        dd::mEdge generateNoiseOperation(dd::mEdge   dd_operation,
+                                         signed char target,
+                                         bool        amplitudeDamping,
+                                         bool        multiQubitOperation) {
+            std::uint_fast8_t effect;
+            for (const auto& noise_type: gateNoiseEffects) {
+                if (noise_type != dd::noiseOperations::amplitudeDamping) {
+                    effect = returnNoiseOperation(noise_type, dist(generator), multiQubitOperation);
+                } else {
+                    if (amplitudeDamping) {
+                        effect = qc::ATrue;
+                    } else {
+                        effect = qc::AFalse;
+                    }
+                }
```

Comment:
all these enum values are not available here as they reside in the QFR. 
I would argue, that this could be reason enough, to move this class to the `include/dd` part of the QFR.

Reply:
Moved it into the qfr